Core/src/main/java/org/tribuo/dataset/SelectedFeatureDataset.java

```diff
@@ -0,0 +1,358 @@
+                copy.removeFeatures(features);
+            }
+            // If the example is still valid then add it, otherwise increment the removed counter
+            if (copy.size() > 0) {
+                data.add(copy);
+                outputInfo.observe(ex.getOutput());
+            } else {
+                tmpNumExamplesRemoved++;
+            }
+        }
+        numExamplesRemoved = tmpNumExamplesRemoved;
+
+        // Rebuild feature and output maps
+        this.featureIDMap = new ImmutableFeatureMap(featureMap);
+        this.outputIDInfo = outputInfo.generateImmutableOutputInfo();
+
+        if (numExamplesRemoved > 0) {
+            logger.info(String.format("filtered out %d examples because they had zero features after the selected feature set was applied.", numExamplesRemoved));
+        }
+    }
+
+    private SelectedFeatureDataset(DataProvenance provenance, OutputFactory<T> factory, String tribuoVersion,
+                                   ImmutableFeatureMap fmap, ImmutableOutputInfo<T> outputInfo,
+                                   List<Example<T>> examples, int k, SelectedFeatureSet featureSet,
+                                   Set<String> selectedFeatures, int numExamplesRemoved) {
+        super(provenance,factory,tribuoVersion,fmap,outputInfo,examples,false);
+        this.k = k;
+        this.selectedFeatures = Collections.unmodifiableSet(selectedFeatures);
+        this.featureSet = featureSet;
+        this.numExamplesRemoved = numExamplesRemoved;
+    }
+
+    /**
+     * Deserialization factory.
+     * @param version The serialized object version.
+     * @param className The class name.
+     * @param message The serialized data.
+     */
+    @SuppressWarnings({"unchecked","rawtypes"}) // guarded & checked by getClass checks.
+    public static SelectedFeatureDataset<?> deserializeFromProto(int version, String className, Any message) throws InvalidProtocolBufferException {
+        if (version < 0 || version > CURRENT_VERSION) {
+            throw new IllegalArgumentException("Unknown version " + version + ", this class supports at most version " + CURRENT_VERSION);
+        }
+        SelectedFeatureDatasetProto proto = message.unpack(SelectedFeatureDatasetProto.class);
+        DatasetDataCarrier<?> carrier = DatasetDataCarrier.deserialize(proto.getMetadata());
+        Class<?> outputClass = carrier.outputFactory().getUnknownOutput().getClass();
+        FeatureMap fmap = carrier.featureDomain();
+        List<Example<?>> examples = new ArrayList<>();
+        for (ExampleProto e : proto.getExamplesList()) {
+            Example<?> example = Example.deserialize(e);
+            if (example.getOutput().getClass().equals(outputClass)) {
+                for (Feature f : example) {
+                    if (fmap.get(f.getName()) == null) {
+                        throw new IllegalStateException("Invalid protobuf, feature domain does not contain feature " + f.getName() + " present in an example");
```

Comment:
Adding the index of the example would make debugging this much easier when thrown